Introduce Installment payment method

guzzilar opened the pull request:
### 1. Objective

Installment payment method is the most wanted feature that hase been asked and requested through multiple channels for Omise-WooCommerce for all the time since Omise started providing Installment feature in Omise core API.

This pull request is aiming to introduce Installment payment method to Omise-WooCommerce plugin that would enable the more selling-channels for merchants and more payment-option for all customers (aka. buyer).

**Related information**:
Related issue(s): #92 "Support for installment payment ?"

### 2. Description of change

Adding a new payment method has always been a big thing that makes tons of line-changes in the code base. So, to break-down the mess, let me walk you through each commit to explain the change in each step.

- 71879d7700497c05dee673ebf66402cf302ea0c9: Implementing a foundation code, adding a minimum code to meet WooCommerce requirement in order to register a new payment method to WooCommerce payment system.
  
  _**Installment payment method** will be listed at `WooCommerce` > `Settings` > `Payments` page._ 👇
  <img width="1552" alt="screen shot 2562-01-24 at 10 49 45" src="https://user-images.githubusercontent.com/2154669/51653398-d25f7380-1fc5-11e9-94de-ed31cb578557.png">
  _Providing with **Omise Installment Setting Page**._ 👇
  <img width="1552" alt="screen shot 2562-01-24 at 10 49 56" src="https://user-images.githubusercontent.com/2154669/51653407-dab7ae80-1fc5-11e9-88a0-e36e4024ca97.png">

- b1d5920b29743579e186d501bf8994564732c8c3: Listing **Installment payment method** at Omise Setting page (for admin).
  <img width="1179" alt="screen shot 2562-01-24 at 10 58 17" src="https://user-images.githubusercontent.com/2154669/51653720-477f7880-1fc7-11e9-8956-0d209d96e384.png">

- 06df14ab2522f2781a9559b461b21e9e4deb3e50: Be able to save the Installment settings to WC configuration table.

- 1e3f67acbe39d8b0abe1b0881ca3f652a9d15a0c Displaying Installment form at the checkout page.
<img width="1552" alt="Screen Shot 2562-04-04 at 12 22 27" src="https://user-images.githubusercontent.com/2154669/55531449-9e46b400-56d4-11e9-8c9a-74db3c246c4f.png">

- 18dcf829d6e39b75c1ba5d18233aef81b026926a Integrating with Omise Capabilities API

- 33b583f0130eed75efbb20f5a8246f1e95815098 Be able to place an order with Installment payment method.

- 5bf4397d7f221627d39b9aebaed5cc3463f24268 Implementing a callback to validate a charge result.

### 3. Quality assurance

**🔧 Environments:**

- **WordPress**:  v5.0.3.
- **WooCommerce**:  v3.3.5.
- **PHP version**: 5.6.30

**✏️ Details:**

**1. Make sure that you can enable Installment payment method properly after installed the plugin.**
At the Omise setting page (wp-admin/admin.php?page=omise-settings), Omise Installment payment method should be listed at the **Payment Methods** section as the following screenshot:
1.1). Click "Config" at the Installment payment method
<img width="1552" alt="installment-01" src="https://user-images.githubusercontent.com/2154669/57132147-dec74980-6dc8-11e9-969d-5d16f5879317.png">

1.2). Check at the "Enable" option, then "save config".
<img width="1552" alt="installment-02" src="https://user-images.githubusercontent.com/2154669/57132228-1f26c780-6dc9-11e9-9275-6a483f284ae5.png">

1.3). Once done, the Installment payment method should be displayed properly at the checkout page with 5 instalment providers.
<img width="1552" alt="installment-03" src="https://user-images.githubusercontent.com/2154669/57132313-61500900-6dc9-11e9-9d8b-776f5abaa0d6.png">

...

**2. Make sure that we can create a new charge using Installment payment method.**
2.1). You may choose any Installment provider from the list at the checkout page, then "Place Order".
<img width="1552" alt="installment-04" src="https://user-images.githubusercontent.com/2154669/57132494-fce17980-6dc9-11e9-8129-cb379b6b4454.png">

2.2). At the test page, click "Success" to mark a charge as successful.
<img width="1552" alt="installment-05" src="https://user-images.githubusercontent.com/2154669/57132617-521d8b00-6dca-11e9-82f7-eb85ce6dd5e0.png">

2.3). You will be redirected back to the store, order confirmation page. Indicating that your payment process has been finished.
<img width="1552" alt="installment-06" src="https://user-images.githubusercontent.com/2154669/57132680-8abd6480-6dca-11e9-84d4-28210774a5d5.png">

2.4). As an admin of the store, you would be able to see a new order has been placed with Installment payment method, status will be marked as `processing`.
<img width="1552" alt="installment-07" src="https://user-images.githubusercontent.com/2154669/57132852-10d9ab00-6dcb-11e9-92df-e8df33b9a344.png">

2.5). At the Omise Dashboard - charge page, there will be a transaction created by the latest purchase that we have placed at the step [2.2].
<img width="1552" alt="installment-08" src="https://user-images.githubusercontent.com/2154669/57133313-96118f80-6dcc-11e9-8e3f-78d629e6854f.png">

...

**3. Make sure that the plugin can handle a failure payment case properly.**
3.1). Repeating the step [2.1] - [2.2] but instead of marking a charge as `success`, we will mark it as `failure` instead. Click `Failure` button.
<img width="1552" alt="installment-09" src="https://user-images.githubusercontent.com/2154669/57135470-b85adb80-6dd3-11e9-8bed-9c3875617bd0.png">

3.2). You will be redirected back to the checkout page with a failure message displaying on a screen.
<img width="1552" alt="installment-10" src="https://user-images.githubusercontent.com/2154669/57135534-eb9d6a80-6dd3-11e9-84c2-defce2a2196c.png">

3.3). As an admin of the store, you would be able to see a new order has been placed with Installment payment method, status will be marked as `failed`.
<img width="1552" alt="installment-11" src="https://user-images.githubusercontent.com/2154669/57135656-5f3f7780-6dd4-11e9-8031-6b28441a107b.png">

**4. Make sure that the plugin can handle a pending payment case properly.**
It is a bit tricky to test this case but absolutely doable.
First of all, you will need to modify the plugin's codebase to fake a charge status from either `successful` or `failure` to `pending`.
4.1). Add a following code at [`/includes/gateway/class-omise-payment-installment.php` after the line: `175`](https://github.com/omise/omise-woocommerce/pull/99/files#diff-aa8a0d75fd53ad3e577f829504c5d5b3R175).
```php
$charge['status']     = 'pending';
$charge['authorized'] = false;
$charge['paid']       = false;
```

4.2). Repeating the step [2.1]-[2.2]. Now you can choose either `success` or `failure`

4.3). You will be redirected back to the store the same as the step [2.3].

4.4). However, the WooCommerce order status will be marked as `on-hold` instead of `processing`. Indicating that the current transaction is in progress and unless you are using the Webhook feature _(see, case 5th)_, merchants will need to recheck with Omise Dashboard later on (could be a couple minutes or a couple hours, depending on an installment provider).
<img width="1552" alt="installment-12" src="https://user-images.githubusercontent.com/2154669/57136161-1ee0f900-6dd6-11e9-9d8f-5431008d6d06.png">

**5. Make sure that the Webhook works properly for Installment payment.**
This test requires for a real server with `https` protocol. However, you may use https://ngrok.com as an alternative.

Also, in order to successfully test this case, you will need to prepare your code the same as [4.1].

5.1). After set up `ngrok` and harded-code a charge result, now go to the Omise setting page on your WordPress Admin (`wp-admin/admin.php?page=omise`). Copy a Webhook endpoint from the section **Webhook endpoint** at the page.
<img width="1552" alt="installment-13" src="https://user-images.githubusercontent.com/2154669/57277955-72519080-70cf-11e9-9566-30fa4289865d.png">

5.2). Paste this endpoint at the Omise Dashboard > Webhook (`https://dashboard.omise.co/[test/live]/webhooks`) page.
<img width="1552" alt="installment-14" src="https://user-images.githubusercontent.com/2154669/57277984-86958d80-70cf-11e9-90ae-923a20be5cea.png">

5.3). Make a new charge, see [2.1] - [2.3].

5.4). Go to WooCommerce admin, order page. Omise-WooCommerce will be able to receive a Webhook event, then update the order status accordingly as it's shown at the **Order notes** section.
<img width="1552" alt="installment-15" src="https://user-images.githubusercontent.com/2154669/57278161-f86dd700-70cf-11e9-8844-208fbd890efe.png">

### 4. Impact of the change

Nothing

### 5. Priority of change

Normal

### 6. Additional Notes

No

guzzilar commented on assets/css/omise-css.css:
```diff
@@ -21,3 +21,92 @@ fieldset.card-exists {
 #new_card_info:checked ~ fieldset.card-exists {
     display: block;
 }
+
```
Note: would be nice to ignore the CSS file for now as I've provided a unifying-css-files already at the pull request https://github.com/omise/omise-woocommerce/pull/113 (currently there are some stylesheets that are duplicated and bad-organize).

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+			$this->has_fields         = true;
+			$this->method_title       = __( 'Omise Installment', 'omise' );
+			$this->method_description = wp_kses(
+				__( 'Accept payment through <strong>Installment</strong> via Omise payment gateway.', 'omise' ),
```
"Accept installment payments via Omise payment gateway"

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+
+			$this->id                 = 'omise_installment';
+			$this->has_fields         = true;
+			$this->method_title       = __( 'Omise Installment', 'omise' );
```
'Omise Installments'

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+				'enabled' => array(
+					'title'   => __( 'Enable/Disable', 'omise' ),
+					'type'    => 'checkbox',
+					'label'   => __( 'Enable Omise Installment Payment', 'omise' ),
```
'Enable Omise Installment Payments'

guzzilar replied:
@jonrandy ✋ wonder why we use `payments` instead of `payment`? Is it because there are multiple installment providers?

jonrandy replied:
because there are always multiple payments when you pay this way, never just one

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+				'title' => array(
+					'title'       => __( 'Title', 'omise' ),
+					'type'        => 'text',
+					'description' => __( 'This controls the title which the user sees during checkout.', 'omise' ),
```
'This controls the title the user sees during checkout.'

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+				'description' => array(
+					'title'       => __( 'Description', 'omise' ),
+					'type'        => 'textarea',
+					'description' => __( 'This controls the description which the user sees during checkout.', 'omise' )
```
'This controls the description the user sees during checkout.'

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( self::STATUS_FAILED == $charge['status'] ) {
```
Curious why you use this 'backwards' style for equality - it's not wrong, but it just doesn't read very well

guzzilar replied:
@jonrandy WordPress Code style guide, they called "Yoda style" 😂

jonrandy replied:
Ok. Odd style this is, but understand I do

jonrandy commented on includes/gateway/class-omise-payment-installment.php:
```diff
@@ -0,0 +1,262 @@
+				);
+
+				$order->update_status( 'failed' );
+
```
Is this the normal way to do a redirect in WC? Feels like we should maybe be calling a utility function that does it

guzzilar replied:
will check on this one (there is wp_redirect() function but not sure why I chose to implement this way at the first place)

guzzilar replied:
<img width="512" alt="Screen Shot 2562-05-13 at 15 51 32" src="https://user-images.githubusercontent.com/2154669/57608391-0ff90400-7597-11e9-8501-9d422d51c296.png">
Juts check WordPress source code. Seems `wp_redirect` is what we should use instead.

Thanks @jonrandy

guzzilar commented:
@jacstn do you have any suggestion for the message "there are no instalment plans..."?

jacstn commented:
@jonrandy ?
Mimimum order value to use installment is 3000 THB?

also.. is it 3k thb? because my order when testing is 3333 THB and it is still not available.

guzzilar commented:
@jacstn Yup It's 3,000. Ah good point, I have to update Omise-PHP lib for Omise-WooCommerce too. This one still use the previous version that was set the minimum installment amount to 5k.

guzzilar commented on includes/class-omise-capabilities.php:
```diff
@@ -0,0 +1,48 @@
+<?php
```
@jacstn so you still prefer me to split the PR? Or this one is ok?

guzzilar commented:
@jacstn updated.
<img width="401" alt="Screen Shot 2562-05-13 at 16 04 42" src="https://user-images.githubusercontent.com/2154669/57609633-7717b800-7599-11e9-958b-cb6b0c0c72a9.png">

guzzilar commented:
Also, if we rebase this PR with the master, you will be able to use Installment payment with an amount 3,000 THB. (I haven't pushed the rebased-branch here, to avoid me having to force push, so I did in my local machine)
<img width="380" alt="Screen Shot 2562-05-13 at 16 10 13" src="https://user-images.githubusercontent.com/2154669/57609735-aaf2dd80-7599-11e9-8d0e-a762db25c50f.png">

guzzilar commented:
@jonrandy @jacstn Hi, can you help re-review again?
I have addressed your comments at:
- https://github.com/omise/omise-woocommerce/pull/99/commits/e2abfff459e7afe5d23e860362eb83d555d7ec35 : Tweaking & correcting English grammar.
- https://github.com/omise/omise-woocommerce/pull/99/commits/910f724b7837ba59311cfdfdfcf5386e1bdb5b11 : Replacing header() function with the WP built-in redirect function.
